File: test/recompute.py
# %%
import numpy as np
import json
from zigffi import _lib
import boto3
import time
import gc
import psycopg2

# %%
# DSN_AIRFLOW = "postgres://son:son@127.0.0.1:6543/costing_airflow4"
# with psycopg2.connect(DSN_AIRFLOW) as airflow_conn:
#     with airflow_conn.cursor() as airflow_cur:
#         DSN_ODOO = "postgres://son:son@127.0.0.1:6543/lion_dev"
#         with psycopg2.connect(DSN_ODOO) as odoo_conn:
#             with odoo_conn.cursor() as odoo_cur:
#                 start_time = time.perf_counter()
#
#                 cs_query = """
#                     SELECT
#                         cs.costing_number,
#                         cs.latest_costing_number_ts,
#                         cs.costing_number_ts
#                     FROM costing_selector cs
#                     WHERE cs.latest_costing_number_ts IS NOT NULL
#                     AND cs.odoo_stt_costing_id IS NOT NULL
#                     AND cs.costing_number_ts != cs.latest_costing_number_ts;
#                 """
#                 airflow_cur.execute(cs_query)
#                 cs_to_update = airflow_cur.fetchall()
#
#                 print(f"\n\nCS_TO_UPDATE:{len(cs_to_update)}\n\n")
#                 latest_cnis = [cs[1] for cs in cs_to_update]
#
#                 costing_number_query = """
#                     SELECT
#                         origin,
#                         destination,
#                         product,
#                         pickup_fee::float,
#                         flight_cost::float,
#                         bucket_file_version,
#                         awb_number,
#                         tuc_number,
#                         to_char(stt_booked_date, 'YYYY-MM-DD HH24:MI:SS') as stt_booked_date,
#                         to_char(stt_pod_date, 'YYYY-MM-DD HH24:MI:SS') as stt_pod_date,
#                         gross_weight::float,
#                         volume_weight::float,
#                         chargeable_weight::float,
#                         client_code,
#                         client_name,
#                         client_category,
#                         route_type,
#                         lag_route_origin,
#                         lag_route_destination,
#                         publish_rate_cost::float,
#                         stt_total_amount::float,
#                         insurance_commission_to_lp::float,
#                         insurance_cost::float,
#                         agent_commision::float,
#                         corporate_discount::float,
#                         woodpacking_fee::float,
#                         pickup_fee_kvp::float,
#                         forward_origin_cost::float,
#                         pcu_fee::float,
#                         outbound_fee::float,
#                         ra_outgoing::float,
#                         wh_outgoing::float,
#                         wh_incoming::float,
#                         truck_cost::float,
#                         train_cost::float,
#                         sea_freight_cost::float,
#                         inbound_fee::float,
#                         delivery_fee_kvp::float,
#                         delivery_fee_mitra::float,
#                         forward_destination_cost_mitra::float,
#                         forward_destination_cost_vendor::float,
#                         cod_commission::float,
#                         lag_route,
#                         lag_moda,
#                         partner_type,
#                         mitra_code_genesis,
#                         poscode_on_going_inv_mitra,
#                         to_char(etl_date, 'YYYY-MM-DD HH24:MI:SS') as etl_date,
#                         handling_cost::float,
#                         other_cost::float,
#                         total_bonus::float,
#                         costing_number_ts,
#                         route_rank
#                     FROM costing_number_ingest
#                     where costing_number_ts = ANY(%(latest_cnis)s)
#                 """
#                 airflow_cur.execute(
#                     costing_number_query,
#                     {"latest_cnis": latest_cnis},
#                 )
#                 latest_cni_values = airflow_cur.fetchall()
#
#                 end_time = time.perf_counter()
#                 elapsed_time = end_time - start_time
#                 print("Elapsed time query cs to update: ", elapsed_time)
#
#                 if latest_cni_values:
#                     start_time = time.perf_counter()
#                     airflow_query, odoo_query = _lib.generate_recompute_queries(
#                         latest_cni_values,
#                     )
#                     end_time = time.perf_counter()
#                     elapsed_time = end_time - start_time
#                     print("Elapsed time process all string: ", elapsed_time)

# %%
latest_cni_values = [
    (
        "PNK",
        "CGK",
        "REGPACK",
        0.0,
        8984.560000000001,
        "costing-bucket/INPUT/lp_costing001.csv_be2c6226-366d-4226-8675-68f958eee10a",
        "990-52686060",
        "None",
        "2024-01-31 02:10:45",
        "2024-02-01 04:48:43",
        3.8,
        0.17,
        4.0,
        "POS2549",
        "PT.ADVENTOUR JELAJAH INDONESIA",
        "RETAIL",
        "INTER-ISLAND",
        "PNK",
        "CGK",
        14700.0,
        84000.0,
        8984.560000000001,
        0.0,
        25200.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        1764.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        "PNK",
        "FLIGHT",
        "KONSOLIDATOR",
        "CONS138",
        "PNK-0230",
        "2024-03-18 07:53:05",
        0.0,
        0.0,
        312.0,
        "20240327T062344;11LP1706667045401;CONS138;1",
        1,
    ),
    (
        "PNK",
        "CGK",
        "REGPACK",
        0.0,
        16777.0,
        "costing-bucket/INPUT/lp_costing001.csv_be2c6226-366d-4226-8675-68f958eee10a",
        "990-52686060",
        None,
        "2024-01-31 02:10:45",
        "2024-02-01 04:48:43",
        3.8,
        0.17,
        4.0,
        "POS2549",
        "POS LION PARCEL KH. AHMAD SYA'YANI",
        "RETAIL",
        "INTER-ISLAND",
        "PNK",
        "CGK",
        14700.0,
        84000.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        12357.6,
        8698.2,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        "PNK",
        "FLIGHT",
        "KONSOLIDATOR",
        "CONS138",
        "PNK-0230",
        "2024-03-18 07:53:05",
        0.0,
        0.0,
        -123.0,
        "20240327T062344;11LP1706667045401;CONS138;3",
        3,
    ),
]

# %%
# n = np.array(latest_cni_values)
# del latest_cni_values
# gc.collect()
# %%
start_time = time.perf_counter()
queries = _lib.generate_recompute_queries(
    latest_cni_values,
)
end_time = time.perf_counter()
elapsed_time = end_time - start_time
print("Elapsed time process all string: ", elapsed_time)
# print(queries)
print(len(queries))
# %%
print("####")
print(queries[0])
print("####")
print(queries[1])

File: src/aggr/recompute.zig
const std = @import("std");

pub const RecomputeFields = struct {
    origin: ?[]const u8 = undefined,
    destination: ?[]const u8 = undefined,
    product: ?[]const u8 = undefined,
    pickup_fee: ?f32 = undefined,
    flight_cost: ?f32 = undefined,
    bucket_file_version: ?[]const u8 = undefined,
    awb_number: ?[]const u8 = undefined,
    tuc_number: ?[]const u8 = undefined,
    stt_booked_date: ?[]const u8 = undefined,
    stt_pod_date: ?[]const u8 = undefined,
    gross_weight: ?f32 = undefined,
    volume_weight: ?f32 = undefined,
    chargeable_weight: ?f32 = undefined,
    client_code: ?[]const u8 = undefined,
    client_name: ?[]const u8 = undefined,
    client_category: ?[]const u8 = undefined,
    route_type: ?[]const u8 = undefined,
    lag_route_origin: ?[]const u8 = undefined,
    lag_route_destination: ?[]const u8 = undefined,
    publish_rate_cost: ?f32 = undefined,
    stt_total_amount: ?f32 = undefined,
    insurance_commission_to_lp: ?f32 = undefined,
    insurance_cost: ?f32 = undefined,
    agent_commision: ?f32 = undefined,
    corporate_discount: ?f32 = undefined,
    woodpacking_fee: ?f32 = undefined,
    pickup_fee_kvp: ?f32 = undefined,
    forward_origin_cost: ?f32 = undefined,
    pcu_fee: ?f32 = undefined,
    outbound_fee: ?f32 = undefined,
    ra_outgoing: ?f32 = undefined,
    wh_outgoing: ?f32 = undefined,
    wh_incoming: ?f32 = undefined,
    truck_cost: ?f32 = undefined,
    train_cost: ?f32 = undefined,
    sea_freight_cost: ?f32 = undefined,
    inbound_fee: ?f32 = undefined,
    delivery_fee_kvp: ?f32 = undefined,
    delivery_fee_mitra: ?f32 = undefined,
    forward_destination_cost_mitra: ?f32 = undefined,
    forward_destination_cost_vendor: ?f32 = undefined,
    cod_commission: ?f32 = undefined,
    lag_route: ?[]const u8 = undefined,
    lag_moda: ?[]const u8 = undefined,
    partner_type: ?[]const u8 = undefined,
    mitra_code_genesis: ?[]const u8 = undefined,
    poscode_on_going_inv_mitra: ?[]const u8 = undefined,
    etl_date: ?[]const u8 = undefined,
    handling_cost: ?f32 = undefined,
    other_cost: ?f32 = undefined,
    bonus_dtpol: ?f32 = undefined,
    costing_number_ts: []const u8 = undefined,
    route_rank: u32 = undefined,

    const Self = @This();

    fn optOrNull(writer: anytype, input: *const RecomputeFields, comptime field_name: []const u8) !void {
        const field_value = @field(input, field_name);
        switch (@TypeOf(field_value)) {
            ?[]const u8 => {
                if (field_value) |content| {

                    // NOTE: replace single quote
                    // 'POS LION PARCEL KH. AHMAD SYA'YANI'
                    // to
                    // 'POS LION PARCEL KH. AHMAD SYA''YANI'
                    //
                    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
                    defer _ = gpa.deinit();
                    const allocator = gpa.allocator();
                    const size = std.mem.replacementSize(u8, content, "'", "''");
                    var output = try allocator.alloc(u8, size);
                    defer allocator.free(output);
                    _ = std.mem.replace(u8, content, "'", "''", output);

                    _ = try writer.print("{s}='{s}',", .{ field_name, output });
                } else {
                    _ = try writer.print("{s}=null,", .{field_name});
                }
            },
            ?f32 => {
                if (field_value) |content| {
                    _ = try writer.print("{s}={d:.2},", .{ field_name, content });
                } else {
                    _ = try writer.print("{s}=null,", .{field_name});
                }
            },
            ?u32 => {
                if (field_value) |content| {
                    _ = try writer.print("{s}={d},", .{ field_name, content });
                } else {
                    _ = try writer.print("{s}=null,", .{field_name});
                }
            },
            else => unreachable,
        }
    }

    pub fn format(
        self: Self,
        comptime fmt: []const u8,
        options: std.fmt.FormatOptions,
        writer: anytype,
    ) !void {
        _ = options;
        if (std.mem.eql(u8, fmt, "odoo")) {
            try self.formatOdoo(writer);
        } else if (std.mem.eql(u8, fmt, "airflow")) {
            try self.formatAirflow(writer);
        }
    }

    fn formatOdoo(c: *const Self, writer: anytype) !void {
        try writer.writeAll("UPDATE stt_middleware_costing SET ");

        try optOrNull(writer, c, "origin");
        try optOrNull(writer, c, "destination");
        try writer.writeAll("write_date=NOW() AT TIME ZONE 'Asia/Jakarta',");
        try optOrNull(writer, c, "product");
        try optOrNull(writer, c, "pickup_fee");
        try optOrNull(writer, c, "flight_cost");

        try optOrNull(writer, c, "bucket_file_version");
        try optOrNull(writer, c, "awb_number");
        try optOrNull(writer, c, "tuc_number");
        try optOrNull(writer, c, "stt_booked_date");
        try optOrNull(writer, c, "stt_pod_date");

        try optOrNull(writer, c, "gross_weight");
        try optOrNull(writer, c, "volume_weight");
        try optOrNull(writer, c, "chargeable_weight");

        try optOrNull(writer, c, "client_code");
        try optOrNull(writer, c, "client_name");
        try optOrNull(writer, c, "client_category");
        try optOrNull(writer, c, "route_type");
        try optOrNull(writer, c, "lag_route_origin");
        try optOrNull(writer, c, "lag_route_destination");

        try optOrNull(writer, c, "publish_rate_cost");
        try optOrNull(writer, c, "stt_total_amount");
        try optOrNull(writer, c, "insurance_commission_to_lp");
        try optOrNull(writer, c, "insurance_cost");
        try optOrNull(writer, c, "agent_commision");
        try optOrNull(writer, c, "corporate_discount");
        try optOrNull(writer, c, "woodpacking_fee");
        try optOrNull(writer, c, "pickup_fee_kvp");
        try optOrNull(writer, c, "forward_origin_cost");
        try optOrNull(writer, c, "pcu_fee");
        try optOrNull(writer, c, "outbound_fee");
        try optOrNull(writer, c, "ra_outgoing");
        try optOrNull(writer, c, "wh_outgoing");
        try optOrNull(writer, c, "wh_incoming");
        try optOrNull(writer, c, "truck_cost");
        try optOrNull(writer, c, "train_cost");
        try optOrNull(writer, c, "sea_freight_cost");
        try optOrNull(writer, c, "inbound_fee");
        try optOrNull(writer, c, "delivery_fee_kvp");
        try optOrNull(writer, c, "delivery_fee_mitra");

        try optOrNull(writer, c, "forward_destination_cost_mitra");
        try optOrNull(writer, c, "forward_destination_cost_vendor");
        try optOrNull(writer, c, "cod_commission");

        try optOrNull(writer, c, "lag_route");
        try optOrNull(writer, c, "lag_moda");

        try optOrNull(writer, c, "partner_type");
        try optOrNull(writer, c, "mitra_code_genesis");
        try optOrNull(writer, c, "poscode_on_going_inv_mitra");
        try optOrNull(writer, c, "etl_date");
        try optOrNull(writer, c, "handling_cost");
        try optOrNull(writer, c, "other_cost");

        try optOrNull(writer, c, "bonus_dtpol");

        _ = try writer.print("costing_number_ts='{s}',", .{c.costing_number_ts});
        _ = try writer.print("route_rank={} ", .{c.route_rank});
        // try optOrNull(writer, c, "route_rank");

        var cnit = std.mem.splitScalar(u8, c.costing_number_ts, ';');
        _ = cnit.first();
        _ = try writer.print("WHERE costing_number = '{s};{s};{s}';", .{ cnit.next().?, cnit.next().?, cnit.next().? });
    }

    fn formatAirflow(c: *const Self, writer: anytype) !void {
        try writer.writeAll("UPDATE costing_selector x ");
        try writer.writeAll("SET costing_number_ts = y.latest_costing_number_ts ");
        try writer.writeAll("FROM (SELECT costing_number,costing_number_ts,");
        try writer.writeAll("latest_costing_number_ts FROM costing_selector ");
        var cnit = std.mem.splitScalar(u8, c.costing_number_ts, ';');
        _ = cnit.first();
        _ = try writer.print("WHERE costing_number='{s};{s};{s}' FOR UPDATE) y ", .{ cnit.next().?, cnit.next().?, cnit.next().? });
        try writer.writeAll("WHERE x.costing_number=y.costing_number;");
    }
};
